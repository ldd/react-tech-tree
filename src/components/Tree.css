.Tree {
  background-color: transparent;
  position: relative;
  display: flex;
  justify-content: space-evenly;
  flex-direction: column;
  height: 400px;
}

@keyframes activate {
  from {
    outline-color: gray;
    stroke: gray;
  }
  to {
    outline-color: gold;
    stroke: gold;
  }
}
.Tree .active {
  animation: 1.5s forwards activate;
}

.Tree-Row {
  display: flex;
  justify-content: space-evenly;
}
