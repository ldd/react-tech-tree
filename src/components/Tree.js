import React, { useState, useRef, useEffect } from "react";
import "./Tree.css";
import { Node } from "./Node";
import { Link } from "./Link";
import { getNodeDOMPositions, simplePathMaker } from "../helpers/svgHelper";

const usePosition = ({ links = [], pathMaker = simplePathMaker }) => {
  // https://stackoverflow.com/a/55996413
  const ref = useRef();
  const [nodePositions, setNodePositions] = useState(links);
  const repositionLinks = () =>
    setNodePositions(
      getNodeDOMPositions(nodePositions, ref.current, pathMaker)
    );

  // https://gist.github.com/gaearon/cb5add26336003ed8c0004c4ba820eae
  useEffect(() => {
    repositionLinks();
    window.addEventListener("resize", repositionLinks);
    return () => window.removeEventListener("resize", repositionLinks);
  }, []);
  return { ref, nodePositions };
};

export const Tree = ({
  id = "tech-tree",
  links = [],
  nodes = [],
  NodeElement = Node,
  LinkElement = Link,
  nodeProps = {},
  linkProps = {}
}) => {
  const { pathMaker, ...otherLinkProps } = linkProps;
  const { ref, nodePositions } = usePosition({ links, pathMaker });
  return (
    <div className="Tree" id={id} ref={ref}>
      <svg className="Tree-Links">
        <TreeLink
          links={links}
          LinkElement={LinkElement}
          linkProps={otherLinkProps}
          nodePositions={nodePositions}
        />
      </svg>
      <TreeNodes
        nodes={nodes}
        NodeElement={NodeElement}
        nodeProps={nodeProps}
      />
    </div>
  );
};

/**
 * React Component made up of NodeElements inside Tree-Rows made from a 2D array of nodes
 */
const TreeNodes = ({ nodes = [], NodeElement = Node, nodeProps = {} }) =>
  nodes.map((siblings, depth) => (
    <div key={`tree-row-${depth}`} className="Tree-Row">
      {siblings.map((props, i) => (
        <NodeElement key={`node-${props.id || i}`} {...nodeProps} {...props} />
      ))}
    </div>
  ));

/**
 * React Component made up of LinkElement made from an array of links augmented by nodePositions information
 */
const TreeLink = ({
  links = [],
  LinkElement = Link,
  linkProps = {},
  nodePositions = []
}) => {
  return links.map(({ from, to, ...props }, i) => {
    const { pathData } = nodePositions[i] || {};
    if (!pathData) return null;

    const id = `${from}-${to}`;
    const allLinkProps = { ...linkProps, ...props };
    return (
      <LinkElement id={id} key={id} pathData={pathData} {...allLinkProps} />
    );
  });
};
