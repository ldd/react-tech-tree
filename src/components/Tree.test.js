// import Enzyme from "enzyme";
// import Adapter from "enzyme-adapter-react-16";
import React from "react";
import renderer from "react-test-renderer";
import { Tree } from "./Tree";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";

// Enzyme.configure({ adapter: new Adapter() });

const getLinkId = ({ from = "", to = "" } = {}) => `${from}-${to}`;

describe("Tree", () => {
  const nodesData = [
    [
      { id: "A", name: "A" },
      { id: "B", name: "B" }
    ]
  ];
  const invalidnodesData = [[{ name: "A" }, { name: "B" }]];
  const links = [{ from: "A", to: "B" }];

  let container;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });
  it("can create empty Tree", () => {
    const component = renderer.create(<Tree />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("can create simple Tree", () => {
    const component = renderer.create(<Tree nodes={nodesData} links={links} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("can create simple Tree with invalid node data", () => {
    const component = renderer.create(<Tree nodes={invalidnodesData} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("can create simple Tree with custom NodeElement", () => {
    const NodeElement = ({ id, name }) => <button id={id}>{name}</button>;
    const component = renderer.create(
      <Tree nodes={nodesData} links={links} NodeElement={NodeElement} />
    );
    expect(component.root.findAllByType(NodeElement)).toHaveLength(2);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("can render simple Tree", () => {
    ReactDOM.render(<Tree nodes={nodesData} links={links} />, container);

    // expect all nodeData to have been rendered with className "Node"
    const domNodes = container.querySelectorAll(".Node");
    const nodeDataIds = nodesData.flat().map(node => node.id);
    expect(domNodes.length).toBe(nodeDataIds.length);
    domNodes.forEach((domNode, i) => {
      expect(domNode.id).toBe(nodeDataIds[i]);
    });

    // expect all links to have been rendered with className "Link"
    const domLinks = container.querySelectorAll(".Link");
    expect(domLinks.length).toBe(links.length);
    domLinks.forEach((domLink, i) => {
      const { from, to } = links[i] || {};
      const expectedId = `${from}-${to}`;
      expect(domLink.id).toBe(expectedId);
    });
  });
  it("can render and activate Nodes", () => {
    ReactDOM.render(<Tree nodes={nodesData} links={links} />, container);
    const domNodes = container.querySelectorAll(".Node");
    const clickedNode = domNodes[0];
    act(() => {
      clickedNode.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    const activeNode = container.querySelector(".Node.active");
    expect(activeNode).toBeDefined();
    expect(activeNode).toEqual(expect.objectContaining({ id: clickedNode.id }));

    const activeLink = container.querySelector(".Link.active");
    const expectedLinkId = getLinkId(links[0]);
    expect(activeLink).toBeDefined();
    expect(activeLink).toEqual(expect.objectContaining({ id: expectedLinkId }));
  });
  it("can render and deactivate Nodes", () => {
    ReactDOM.render(<Tree nodes={nodesData} links={links} />, container);
    const domNodes = container.querySelectorAll(".Node");
    const clickedNode = domNodes[0];
    act(() => {
      clickedNode.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    act(() => {
      clickedNode.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    const activeNode = container.querySelector(".Node.active");
    expect(activeNode).toBeNull();

    const activeLink = container.querySelector(".Link.active");
    expect(activeLink).toBeNull();
  });
  it("can render and ignore invalid Links", () => {
    const invalidLinks = [{ from: "A" }];
    ReactDOM.render(<Tree nodes={nodesData} links={invalidLinks} />, container);
    const domNodes = container.querySelectorAll(".Node");
    const clickedNode = domNodes[0];
    act(() => {
      clickedNode.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });
    const activeNode = container.querySelector(".Node.active");
    expect(activeNode).toBeDefined();

    const activeLink = container.querySelector(".Link.active");
    expect(activeLink).toBeNull();
  });
  it("can render with invalid node data", () => {
    ReactDOM.render(<Tree nodes={invalidnodesData} links={links} />, container);
    const domNode = container.querySelector(".Node");
    expect(domNode).toBeDefined();
    const domLink = container.querySelector(".Links");
    expect(domLink).toBeNull();
  });
});
