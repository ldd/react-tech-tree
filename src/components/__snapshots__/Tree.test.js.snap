// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Component [Tree] can create empty Tree 1`] = `
<div
  className="Tree"
  id="tech-tree"
>
  <svg
    className="Link-Container"
  />
</div>
`;

exports[`React Component [Tree] can create simple Tree 1`] = `
<div
  className="Tree"
  id="tech-tree"
>
  <svg
    className="Link-Container"
  />
  <div
    className="Tree-Row"
  >
    <div
      className="Node Sprite"
      id="A"
      onClick={[Function]}
      style={Object {}}
    />
    <div
      className="Node Sprite"
      id="B"
      onClick={[Function]}
      style={Object {}}
    />
  </div>
</div>
`;

exports[`React Component [Tree] can create simple Tree with custom NodeElement 1`] = `
<div
  className="Tree"
  id="tech-tree"
>
  <svg
    className="Link-Container"
  />
  <div
    className="Tree-Row"
  >
    <button
      id="A"
    >
      A
    </button>
    <button
      id="B"
    >
      B
    </button>
  </div>
</div>
`;

exports[`React Component [Tree] can create simple Tree with invalid node data 1`] = `
<div
  className="Tree"
  id="tech-tree"
>
  <svg
    className="Link-Container"
  />
  <div
    className="Tree-Row"
  >
    <div
      className="Node Sprite"
      onClick={[Function]}
      style={Object {}}
    />
    <div
      className="Node Sprite"
      onClick={[Function]}
      style={Object {}}
    />
  </div>
</div>
`;
